README.rst
************************************
phancap - website screenshot service
************************************

Web service (API) to create website screenshots.

Self-hosted and written in PHP. Caching included.


*phancap* is useful for:

- Show screenshots for websites in your bookmarking application
- Archive a HTML page as PDF for later viewing


.. contents::


========
Features
========

* Configurable browser size
* Configurable screenshot size
* Clip and full page rendering (full height)
* JPG, PNG and PDF output (PDFs are searchable)
* Authentication
* Can run on a normal web server without GUI. See dependencies_.
* Generated images get meta data embedded:

  * URL of captured page
  * Capture settings


.. note::
    *phancap* does not rely on a "real" browser.
    Currently ``cutycapt`` is utilized, which uses a pretty bare webkit to render
    the pages.
    Do not expect pixel-for-pixel identical rendering as your desktop browser.


===============
Getting started
===============

Basic setup
===========
#. Download the ``.phar`` file and put it onto your web server
#. Open the phar file in your browser

   If you only see text beginning with ``<?php``, you need to
   `setup .phar file extension handling`__ in your web server first.
#. Click the "setup check" link
#. Fix all errors that are reported
#. Run ``phancap.phar/get.php?url=cweiske.de`` and see the screenshot

__ http://cweiske.de/tagebuch/phar-webserver.htm


Advanced setup
==============
With the basic setup, everyone may use your server to create website
screenshots.
You may want to change that or simply change some default settings.

#. Create a config file ``phancap.phar.config.php``
#. Edit it; see the configuration_ options.


==============
URL parameters
==============
``get.php`` supports the following parameters:

Browser parameters
==================
``url``
  Website URL
``bwidth``
  Browser width (default: 1024)
``bheight``
  Browser height (default: none)

Screenshot parameters
=====================
``swidth``
  Screenshot width (default: none (no scaling))
``sheight``
  Screenshot height (default: none)
``sformat``
  Screenshot format (``png``, ``jpg``, ``pdf``, default: ``png``)
``smode``
  Screenshot mode (``screen`` (4:3) or ``page`` (full website height))
``smaxage``
  Maximum age of screenshot in seconds.
  ISO 8601 duration specifications accepted:

  - ``P1Y`` - 1 year
  - ``P2W`` - 2 weeks
  - ``P1D`` - 1 day
  - ``PT4H`` - 4 hours

  The configuration file defines a minimum age that the user cannot undercut
  (``$screenshotMinAge``), as well as a default value (``$screenshotMaxAge``).

Authentication parameters
=========================
``atimestamp``
  Time at which the request URL was generated (unix timestamp)
``atoken``
  Access token (username)
``asignature``
  Signature for the request. See the authentication_ section.


=============
Configuration
=============
phancap looks at several places for its configuration file:

#. ``phancap.phar.config.php`` in the same directory as your
   ``phancap.phar`` file.

#. ``/etc/phancap.php``


Configuration variables
=======================
``$cacheDir``
  Full file system path to image cache directory
``$cacheDirUrl``
  Full URL to cache directory
``$access``
  Credentials for access control

  ``true`` to allow access to anyone, ``false`` to disable it completely.
  ``array`` of username - secret key combinations otherwise.
``$cutycapt['parameters']``
  Additional command line parameters for cutycapt.
  Can be used to e.g. enable browser plugins:

  ``$cutycapt['parameters'] = '--plugins=on';``
``$cutycapt['maxWaitTime']``
  Maximal time in seconds to wait for cutycapt to finish rendering.
  Defaults to 30 seconds.
``$disableSetup``
  Disable ``setup.php`` which will leak file system paths
``$redirect``
  Redirect to static image urls after generating them
``$timestampmaxAge``
  How long a signature timestamp is considered valid. 2 days default.
``$screenshotMaxAge``
  Cache time of downloaded screenshots.

  When the file is as older than this, it gets re-created.
``$screenshotMinAge``
  Minimum age of a screeshot. 1 hour default.
 
  A user cannot set the max age parameter below it.



==============
Authentication
==============
Creating screenshots of websites is a resource intensive process.
To prevent unauthorized access to the service, phancap supports authentication
via a signature parameter similar to OAuth's ``oauth_signature``.

Phancap's configuration file may contain a ``$access`` variable:

``true``
  Everyone is allowed to access the service
``false``
  Nobody is allowed to access the service
``array``
  A list of usernames that are allowed to request screenshots, together
  with their secret keys (password)::

    $access = array(
       'user1' => 'secret1',
       'user2' => 'secret2',
    )

The signature algorithm is as follows:

#. Parameters ``atimestamp`` (current unix timestamp) and
   ``atoken`` (username) have to be added to the URL parameters

#. URL parameters are normalized as described in
   `OAuth Parameters Normalization`__:

   #. Sort parameters list by name
   #. Name and value are `raw-url-encoded`__
   #. Name and value are concatenated with ``=`` as separator
   #. The resulting strings are concatenated with ``&`` as separator

#. URL parameter string is used together with the secret key
   to create a `HMAC-SHA1`__ digest

#. Digest is appended to the URL as ``asignature``

__ http://tools.ietf.org/html/rfc5849#section-3.4.1.3.2
__ http://tools.ietf.org/html/rfc5849#section-3.6
__ http://tools.ietf.org/html/rfc5849#section-3.4.2


Example
=======

.. note::

    The ``docs/`` directory contains an example PHP client implementation.

We want to create a screenshot of ``http://example.org/`` in size 400x300,
using the browser size of 1024x768::

    http://example.org/phancap/get.php?swidth=400&sheight=300&url=http%3A%2F%2Fexample.org%2F&bwidth=1024&bheight=768

Phancap's config file contains::

    $access = array(
        'user' => 'secret'
    );

Our parameters are thus:

============== =====
Name           Value
============== =====
``swidth``     ``400``
``sheight``    ``300``
``url``        ``http://example.org/``
``bwidth``     ``1024``
``bheight``    ``768``
============== =====

At first, we need to add parameters ``atimestamp`` and ``atoken``.
``atimestamp`` is the current unix timestamp.
``atoken`` is our user name: ``user``.

Now the parameter list is sorted:

============== =====
Name           Value
============== =====
``atimestamp`` ``1396353987``
``atoken``     ``user``
``bheight``    ``768``
``bwidth``     ``1024``
``sheight``    ``300``
``swidth``     ``400``
``url``        ``http://example.org/``
============== =====

The parameters are raw-url-encoded. The only value that changes is the url,
it becomes ``http%3A%2F%2Fexample.org%2F``.

Concatenating the name/value pairs leads to the following string::

    atimestamp=1396353987&atoken=user&bheight=768&bwidth=1024&sheight=300&swidth=400&url=http%3A%2F%2Fexample.org%2F

Creating the HMAC digest with sha1, the calculated string and our key
``secret`` gives us the following string::

    9a12eac5ff859f9306eaaf5a18b9a931fe10b89d

This is the signature; it gets appended to the URL as ``asignature`` parameter.


============
Dependencies
============
- External tools:

  - `cutycapt <http://cutycapt.sourceforge.net/>`_
  - `exiftool <http://owl.phy.queensu.ca/~phil/exiftool/>`_
  - `imagemagick's <http://www.imagemagick.org/>`_ ``convert``
  - ``xvfb-run``

- Libraries (already included in the ``.phar``):

  - PEAR's ``System.php``


=============
About phancap
=============

License
=======
``phancap`` is licensed under the `AGPL v3`__ or later.

__ http://www.gnu.org/licenses/agpl.html


Homepage
========
Web site
   http://cweiske.de/phancap.htm

Source code
   http://git.cweiske.de/phancap.git

   Mirror: https://github.com/cweiske/phancap


Author
======
Written by Christian Weiske, cweiske@cweiske.de


============
Alternatives
============
All of those are open source:

* http://code.google.com/p/browsershots/ (python)
* https://github.com/gre/screenshot-webservice (scala)


===========
Development
===========

Releasing a new version
=======================

#. Update ``ChangeLog``
#. Change version number in ``build.xml``
#. Run ``phing``
#. Deploy the new files in ``dist/``
#. Tag the new version in git


ChangeLog
2016-07-12  Christian Weiske  <cweiske@cweiske.de>

	* Add support for internationalized domain names (IDNA)
	* Add meta data into generated images
	* Fix bug with cutycapt timeouts
	* Release version 0.4.0

2016-06-29  Christian Weiske  <cweiske@cweiske.de>

	* Add form to make a complex request from your browser
	* Prevent infinite loop when phancap tries to capture itself
	* Fix .pdf MIME type
	* Improve HTTP headers when sending out images
	* Improve parameter validation, allow empty parameter values
	* Release version 0.3.0

2014-11-06  Christian Weiske  <cweiske@cweiske.de>

	* Kill cutycapt after some time
	* Make cutycapt options configurable
	* Release version 0.2.0

2014-04-18  Christian Weiske  <cweiske@cweiske.de>

	* Release version 0.1.0


src/phancap/Options.php
<?php
/**
 * Part of phancap
 *
 * PHP version 5
 *
 * @category  Tools
 * @package   Options
 * @author    Christian Weiske <cweiske@cweiske.de>
 * @copyright 2014 Christian Weiske
 * @license   http://www.gnu.org/licenses/agpl.html GNU AGPL v3
 * @link      http://cweiske.de/phancap.htm
 */
namespace phancap;

/**
 * Options a user can give to the API
 *
 * @category  Tools
 * @package   Options
 * @author    Christian Weiske <cweiske@cweiske.de>
 * @copyright 2014 Christian Weiske
 * @license   http://www.gnu.org/licenses/agpl.html GNU AGPL v3
 * @version   Release: @package_version@
 * @link      http://cweiske.de/phancap.htm
 */
class Options
{
    /**
     * Available options and their configuration
     *
     * @var array
     */
    public $options = array(
        /**
         * Browser settings
         */
        'url' => array(
            'title'     => 'Website URL',
            'default'   => null,
            'type'      => 'url',
            'required'  => true,
        ),
        'bwidth' => array(
            'title'   => 'Browser width',
            'default' => 1024,
            'type'    => 'int',
            'min'     => 16,
            'max'     => 2048,
        ),
        'bheight' => array(
            'title'   => 'Browser height',
            'default' => null,
            'type'    => 'int',
            'min'     => 16,
            'max'     => 8192,
        ),
        /**
         * Screenshot settings
         */
        'swidth' => array(
            'title'   => 'Screenshot width',
            'default' => null,
            'type'    => 'int',
            'min'     => 16,
            'max'     => 8192,
        ),
        'sheight' => array(
            'title'   => 'Screenshot height',
            'default' => null,
            'type'    => 'int',
            'min'     => 16,
            'max'     => 8192,
        ),
        'sformat' => array(
            'title'   => 'Screenshot format',
            'default' => 'png',
            'type'    => array('png', 'jpg', 'pdf'),
        ),
        'smode' => array(
            'title'   => 'Screenshot mode',
            'default' => 'screen',
            'type'    => array('screen', 'page'),
        ),
        'smaxage' => array(
            'title'   => 'Maximum age for a screenshot',
            'default' => null,
            'type'    => 'age',
            'min'     => null,
        ),
        /**
         * Authentication
         */
        'atimestamp' => array(
            'title'   => 'Timestamp the request has been generated',
            'default' => null,
            'type'    => 'skip',
        ),
        'atoken' => array(
            'title'   => 'Access token (user name)',
            'default' => null,
            'type'    => 'skip',
        ),
        'asignature' => array(
            'title'   => 'Access signature',
            'default' => null,
            'type'    => 'skip',
        ),
    );

    /**
     * Actual values we use after parsing the GET parameters
     *
     * @var array
     */
    public $values = array();

    /**
     * Configuration object
     *
     * @var Config
     */
    protected $config;


    /**
     * Parses an array of options, validates them and writes them into
     * $this->values.
     *
     * @param array $arValues Array of options, e.g. $_GET
     *
     * @return void
     * @throws \InvalidArgumentException When required parameters are missing
     *         or parameter values are invalid.
     */
    public function parse($arValues)
    {
        foreach ($this->options as $name => $arOption) {
            $this->values[$name] = $arOption['default'];
            if (!isset($arValues[$name])) {
                if (isset($arOption['required'])) {
                    throw new \InvalidArgumentException(
                        $name . ' parameter missing'
                    );
                }
                continue;
            }

            if ($arValues[$name] === ''
                && !isset($arOption['required'])
            ) {
                //allow empty value; default value will be used
            } else if ($arOption['type'] == 'url') {
                $this->values[$name] = $this->validateUrl($arValues[$name]);
            } else if ($arOption['type'] == 'int') {
                $this->values[$name] = $this->validateInt(
                    $name, $arValues[$name],
                    $arOption['min'], $arOption['max']
                );
            } else if (gettype($arOption['type']) == 'array') {
                $this->values[$name] = $this->validateArray(
                    $name, $arValues[$name], $arOption['type']
                );
            } else if ($arOption['type'] == 'age') {
                $this->values[$name] = $this->clamp(
                    static::validateAge($arValues[$name]),
                    $arOption['min'], null,
                    true
                );
            } else if ($arOption['type'] != 'skip') {
                throw new \InvalidArgumentException(
                    'Unsupported option type: ' . $arOption['type']
                );
            }
            unset($arValues[$name]);
        }

        if (count($arValues) > 0) {
            throw new \InvalidArgumentException(
                'Unsupported parameter: ' . implode(', ', array_keys($arValues))
            );
        }

        $this->calcPageSize();
    }

    /**
     * Calculate the browser size and screenshot size from the given options
     *
     * @return void
     */
    protected function calcPageSize()
    {
        if ($this->values['swidth'] === null) {
            $this->values['swidth'] = $this->values['bwidth'];
        }
        if ($this->values['smode'] == 'page') {
            return;
        }

        if ($this->values['sheight'] !== null) {
            $this->values['bheight'] = intval(
                $this->values['bwidth'] / $this->values['swidth']
                * $this->values['sheight']
            );
        } else if ($this->values['bheight'] !== null) {
            $this->values['sheight'] = intval(
                $this->values['swidth'] / $this->values['bwidth']
                * $this->values['bheight']
            );
        } else {
            //no height set. use 4:3
            $this->values['sheight'] = $this->values['swidth'] / 4 * 3;
            $this->values['bheight'] = $this->values['bwidth'] / 4 * 3;
        }
    }

    /**
     * Makes sure a value is between $min and $max (inclusive)
     *
     * @param integer $value  Value to check
     * @param integer $min    Minimum allowed value
     * @param integer $max    Maximum allowed value
     * @param boolean $silent When silent, invalid values are corrected.
     *                        An exception is thrown otherwise.
     *
     * @return integer Corrected value
     * @throws \InvalidArgumentException When not silent and value outside range
     */
    protected function clamp($value, $min, $max, $silent = false)
    {
        if ($min !== null && $value < $min) {
            if ($silent) {
                $value = $min;
            } else {
                throw new \InvalidArgumentException(
                    'Value must be at least ' . $min
                );
            }
        }
        if ($max !== null && $value > $max) {
            if ($silent) {
                $value = $max;
            } else {
                throw new \InvalidArgumentException(
                    'Value may be up to ' . $min
                );
            }
        }
        return $value;
    }

    /**
     * Validates an age is numeric. If it is not numeric, it's interpreted as
     * a ISO 8601 duration specification.
     *
     * @param string $value Age in seconds
     *
     * @return integer Age in seconds
     * @throws \InvalidArgumentException
     * @link   http://en.wikipedia.org/wiki/Iso8601#Durations
     */
    public static function validateAge($value)
    {
        if (!is_numeric($value)) {
            //convert short notation to seconds
            $value = 'P' . ltrim(strtoupper($value), 'P');
            try {
                $interval = new \DateInterval($value);
            } catch (\Exception $e) {
                throw new \InvalidArgumentException(
                    'Invalid age: ' . $value
                );
            }
            $value = 86400 * (
                $interval->y * 365
                + $interval->m * 30
                + $interval->d
            ) + $interval->h * 3600
                + $interval->m * 60
                + $interval->s;
        }
        return $value;
    }

    /**
     * Check that a given value exists in an array
     *
     * @param string $name    Variable name
     * @param string $value   Value to check
     * @param array  $options Array of allowed values
     *
     * @return string Value
     * @throws \InvalidArgumentException If the value does not exist in $options
     */
    protected function validateArray($name, $value, $options)
    {
        if (array_search($value, $options) === false) {
            throw new \InvalidArgumentException(
                'Invalid value ' . $value . ' for ' . $name . '.'
                . ' Allowed: ' . implode(', ', $options)
            );
        }
        return $value;
    }

    /**
     * Validate that a value is numeric and between $min and $max (inclusive)
     *
     * @param string  $name  Variable name
     * @param string  $value Value to check
     * @param integer $min   Minimum allowed value
     * @param integer $max   Maximum allowed value
     *
     * @return integer Value as integer
     * @throws \InvalidArgumentException When outside range or not numeric
     */
    protected function validateInt($name, $value, $min, $max)
    {
        if (!is_numeric($value)) {
            throw new \InvalidArgumentException(
                $name . ' value must be a number'
            );
        }
        $value = (int) $value;
        return $this->clamp($value, $min, $max);
    }

    /**
     * Validate (and fix) an URL
     *
     * @param string $url URL
     *
     * @return string Fixed URL
     * @throws \InvalidArgumentException
     */
    protected function validateUrl($url)
    {
        if ($url === '') {
            throw new \InvalidArgumentException('URL is empty');
        }
        $parts = parse_url($url);
        if ($parts === false) {
            throw new \InvalidArgumentException('Invalid URL');
        }
        if (!isset($parts['scheme'])) {
            $url = 'http://' . $url;
            $parts = parse_url($url);
        } else if ($parts['scheme'] != 'http' && $parts['scheme'] != 'https') {
            throw new \InvalidArgumentException('Unsupported protocol');
        }
        if (!isset($parts['host'])) {
            throw new \InvalidArgumentException('URL host missing');
        }

        $rebuild = false;
        if (strlen(preg_replace('#[[:ascii:]]#', '', $parts['host']))) {
            //non-ascii characters in the host name
            $host = idn_to_ascii($parts['host']);
            if ($host === false) {
                //incoming URL was not UTF-8 but some ISO dialect
                $host = idn_to_ascii(utf8_encode($parts['host']));
                if ($host === false) {
                    throw new \InvalidArgumentException(
                        'Strange characters in host name'
                    );
                }
            }
            $parts['host'] = $host;
            $rebuild = true;
        }
        if (strlen(preg_replace('#[[:ascii:]]#', '', $parts['path']))) {
            //non-ascii characters in the path
            $parts['path'] = str_replace('%2F', '/', urlencode($parts['path']));
            $rebuild = true;
        }

        if ($rebuild) {
            $url = static::http_build_url($parts);
        }

        return $url;
    }

    /**
     * Set phancap configuration
     *
     * @param Config $config Phancap configuration
     *
     * @return void
     */
    public function setConfig(Config $config)
    {
        $this->config = $config;
        $this->options['smaxage']['default'] = $this->config->screenshotMaxAge;
        $this->options['smaxage']['min']     = $this->config->screenshotMinAge;
    }

    /**
     * Re-build an URL parts array generated by parse_url()
     *
     * @param string $parts Array of URL parts
     *
     * @return string URL
     */
    protected static function http_build_url($parts)
    {
        $scheme   = isset($parts['scheme'])
            ? $parts['scheme'] . '://' : '';
        $host     = isset($parts['host'])
            ? $parts['host'] : '';
        $port     = isset($parts['port'])
            ? ':' . $parts['port'] : '';
        $user     = isset($parts['user'])
            ? $parts['user'] : '';
        $pass     = isset($parts['pass'])

            ? ':' . $parts['pass']  : '';
        $pass     = ($user || $pass)
            ? "$pass@" : '';
        $path     = isset($parts['path'])
            ? $parts['path'] : '';
        $query    = isset($parts['query'])
            ? '?' . $parts['query'] : '';
        $fragment = isset($parts['fragment'])
            ? '#' . $parts['fragment'] : '';
        return "$scheme$user$pass$host$port$path$query$fragment";
    }
}
?>
